### PR TITLE
Add static method ActiveRecord::saveMultiple()

### PR DESCRIPTION
**Pull request by WinterSilence**
Added to ActiveRecord because unable to create intermediate class StoreableModel between Model and BaseActiveRecord

| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ❌
| Fixed issues  |

### CODE REVIEW

**Comment by bizley** [comment]
What value does it add to the core? This is just the shortcut and you can easily do it in your own app.

**Comment by WinterSilence** [comment]
@bizley you can say this about half methods of AR

**Comment by bizley** [comment]
I disagree and I would reject this addition but since I'm not a decision maker here I would like to hear opinion of others.

**Comment by WinterSilence** [comment]
@bizley disagree with what? For example, look [validateMultiple()](https://github.com/yiisoft/yii2/blob/master/framework/base/Model.php#L930), doesn't remind you of anything?

**Comment by bizley** [comment]
I disagree about "half of methods" but yes, this example is a valid point.

**Comment by WinterSilence** [comment]
@bizley what's about getDb(), findBySql(), updateAll(), deleteAll(), find(), getTableSchema(), primaryKey(), attributes(), equals(), isTransactional()? ))

**Comment by terabytesoftw** [comment]
This type of function should not be inside the core, there are different approaches for handling it, so it should be part of the user's code.

**Comment by WinterSilence** [comment]
@terabytesoftw as `validateMultiple()`? (: 

> there are different approaches for handling it

for example?

**Comment by machour** [comment]
I don't see any added value, and if one of the models isn't saved, we won't know which one it is.
Better to leave this out of the framework.

**Comment by WinterSilence** [comment]
@machour
> we won't know which one it is

if method return false, then you can display form with errors by ActiveForm, see [manual example](https://www.yiiframework.com/doc/guide/2.0/en/input-tabular-input#updating-a-fixed-set-of-records)

**Comment by schmunk42** [comment]
As said before :) Yii 2 is actually in a feature freeze state.
It's a general questions if we want to add new `public` methods, IMHO this should be *very* limited.

And if we do so, we'd need extensive tests for it.

**Comment by WinterSilence** [comment]
@schmunk42 i don't try to add new features, I only complete current features missed in current implementation. adding tests in PR it's not problem.

**Comment by schmunk42** [comment]
> added method missed in current implementation

In which way is this "missed"? 

There's no implementation under the name `saveMultiple` somewhere else, just discussions.
https://github.com/yiisoft/yii2/search?q=saveMultiple&type=issues

**Comment by WinterSilence** [comment]
@schmunk42 if you can load and validate multiple model, then there must be a way to save them too. https://www.yiiframework.com/doc/guide/2.0/en/input-tabular-input#updating-a-fixed-set-of-records look like "white spot" in model API

**Comment by schmunk42** [comment]
Related discussions:
- https://github.com/yiisoft/yii2/issues/7585
- https://github.com/yiisoft/yii2/issues/8565
- https://github.com/yiisoft/yii2/issues/8620

**Comment by WinterSilence** [comment]
@schmunk42 it does not explain why loadMultiple/validateMultiple is normal, but not saveMultiple.

**Comment by darkdef** [comment]
Sorry, but me don't like this enhancement.
Also i'm think updateAll, deleteAll - wrong examples. It's commands having alternatives in sql queries. 
saveMultiple - it's example of typical business logic and this shouldn't be parts of framework. As well as the validateMultiple method. But removing it would be BC

**Comment by WinterSilence** [comment]
@darkdef
> saveMultiple - it's example of typical business logic and this shouldn't be parts of framework.

`*Multiple` methods are abstract short-hands and not contains domain/business logic

> Also i'm think updateAll, deleteAll - wrong examples. It's commands having alternatives in sql queries.

don't understand why they are wrong examples. `updateAll()` not contain extra logic, it's inline short-hand - you can call `static::getDb()->createCommand()->update(static::tableName(), $attributes, $condition, $params)->execute();` directly

> this shouldn't be parts of framework
> As well as the validateMultiple method

they can be part of framework, but in other place (trait, model's collection)

>  But removing it would be BC

Is not abstract conversation about whether it's right or wrong way, conversation about whether it's right for current implementation: if class have `loadMultiple()` and `validateMultiple()`, then `saveMultiple()` look like logical solution.

'*Multiple' methods added to package by his owners (as you). Owners(you) could have moved these methods into separate behaviour/trait for long time, but they(you) didn't do this.

**Comment by rustamwin** [comment]
Note that the `validateMultiple`, `loadMultiple` methods are in the Model class, not in ActiveRecord.

**Comment by WinterSilence** [comment]
@rustamwin yes captain, but `save()` declared in `BaseActiveRecord`. Why `saveMultiple()` declared not in `BaseActiveRecord`? To replacing `ActiveRecord` in application (see Base* class pattern). 
Current implementation `ActiveRecord` declare/override methods i.e. violates Base* class pattern(not for discuss here)

File: framework/db/ActiveRecord.php [hunk 1]
```diff
@@ -800,4 +800,26 @@ public function isTransactional($operation)
...
+     * @return bool returns true, if all models successfully saved, returns false, if some models not saved (check by [[hasError()]])
+     * @see save()
+     * @since 2.0.46
+     */
+    public static function saveMultiple($models, $runValidation = true, $attributeNames = null)
+    {
+        $result = true;
+        /** @var ActiveRecord $model */
+        foreach ($models as $model) {
+            $result = $model->save($runValidation || $model->hasErrors(), $attributeNames) && $result;
```
Comment (darkdef) [inline framework/db/ActiveRecord.php:820]: How to determine which model has a validation (save) error and what is it?

Comment (WinterSilence) [reply framework/db/ActiveRecord.php:820]: @darkdef as before: if method return true, action redirect to page "list/index"(CRUD), else, display form with errors

**Comment by WinterSilence** [comment]
@bizley @terabytesoftw @machour @schmunk42 @darkdef Do you have any objections? In Yii2 context and  not personal. Let's try give answer: why `loadMultiple()`/`validateMultiple()` is normal, but not `saveMultiple()`?
P.S. look to this as visit to doctor: nobody like it (doctor too), but it's necessary to normal work.

**Comment by machour** [comment]
I do have an objection, that I mentioned and you discarded, so I don't know why you're asking again 🤷🏼 
And Yii2 is supposed to be in feature-freeze mode.

**Comment by WinterSilence** [comment]
@machour 
> In Yii2 context and not personal

> I do have an objection, that I mentioned and you discarded

your objection NOT in Yii2 context and don't give answer on question:
 
> why loadMultiple()/validateMultiple() is normal, but not saveMultiple()?

**Comment by machour** [comment]
> And Yii2 is supposed to be in feature-freeze mode. 

Because they were there before the feature freeze ?
Anyway, please don't mention me anymore to debate this :)

**Comment by darkdef** [comment]
It was wrong to add loadMultiple/validateMultiple to  ActiveRecord domain.  But now their removal is impossible without bc.  My vote is against this change.
But It's great idea for refactoring of Yii3 active-record. Moving all of that's methods to separate class

**Comment by bizley** [comment]
Ok, looks like most of the team is against adding this method so we will not do it. Thank you all for your time and the input.

**Comment by WinterSilence** [comment]
@darkdef 
> But It's great idea for refactoring of Yii3 active-record. Moving all of that's methods to separate class

We can't do this, it's possible only in next major version

@machour the project is frozen when no one wants to support/update them. in current case it's not true. if you don't want change yii2, don't do it, but don't stop others from doing it.

@bizley reason to close?

**Comment by machour** [comment]
You are absolutely wrong. 

The project is on feature-freeze, not because I don't want to change it, but by a team decision:
  https://www.yiiframework.com/news/156/yii-2-1-and-yii-2-0-feature-freeze

Read Yii 3 instead of Yii 2.1 in the announcement. 
It was decided to freeze feature so that the team can focus on the next version.

You are very welcome to submit new features to Yii 3.

You are also free to start a discussion about wether or not we un-freeze Yii 2, before bullying people into merging your PRs.

You are also free to fork the Yii2 framework and maintain your own fork with your own policy.

**Pull request closed without merging**